jgrapht-core/src/test/java/org/jgrapht/traverse/DepthFirstIteratorTest.java

```diff
@@ -55,7 +60,7 @@ String getExpectedStr2()
     @Override
     String getExpectedFinishString()
     {
-        return "6:4:9:2:8:7:5:3:1:orphan:";
+        return "1:3:6:5:7:9:4:8:2:orphan:";
```

Comment:
You've eliminated the old white/gray/black bookkeeping that was producing the correct finish order.  Why?

Reply:
I'm not entirely sure that I understand your comment. In DepthFirstSearch there is no such thing as a 'correct finish order'. Quoting from the original author of the test class:

> NOTE: This test uses hard-coded expected ordering isn't really guaranteed by the specification of
> the algorithm. This could cause false failures if the traversal implementation changes.
In other words, alternative Depth First Search implementations may produce a different traversal order, while still producing a different but correct Depth First Search Tree.

The original implementation using white/gray/black bookkeeping was surprisingly slow. Some tests on reasonable sized graphs gave me one to two orders of magnitude speedup with the new implementation. Unless I misunderstood something, the book keeping was rather unnecessary.

Reply:
It's true that the finish order is not deterministic, which I believe is what that comment meant.  However, there are constraints on the allowable orderings:

http://www.personal.kent.edu/~rmuhamma/Algorithms/MyAlgorithms/GraphAlgor/depthSearch.htm

So I don't think we can eliminate the existing logic.  I'm guessing the dominant cost is in finding+removing nodes from the middle of the deque, so a better data structure is called for if we want to optimize that.

(If that's not possible, then instead of replacing the existing iterator, we could just add a new one.)

Reply:
I studied the reference you stated - the coloring logic being used (white/grey/black) is just syntactic sugar which simplifies the discussion of the algorithm, as well as some of the proofs. The implementation in this PR satisfies all constraints wrt allowable orderings, while being significantly more efficient. It's quite easy to see how different (but valid) orderings can be obtained.

- Example 1: triangle graph (a,b)(b,c)(c,a). A valid DFS ordering: [a,b,c]. Another valid ordering: [a,c,b].
- Example 2: even without cycles, different orderings exist. For instance, in a graph with 2 edge (a,b)(a,c) again two orderings exist: [a,b,c] and [a,c,b].
The existence of different orderings is a direct consequence of the fact that in a graph, the neighbors of a vertex form an unordered set.

Just to be sure, I also checked some other reference works:

- Graphs Networks and algorithms by Dieter Jungnickel
- Introduction to Algorithms by Thomas H. Cormen , Charles E. Leiserson, Ronald L. Rivest, Clifford Stein
None of these references state additional constraints on feasible orderings aside from the ones that have already been implemented in this PR.

Reply:
I'm referring to the constraint between discovery order and finish order.  See the "parenthesis theorem" in the link above.  Your getExpectedFinishString is identical to getExpectedStr2 because you are calling finishVertex as soon as you encounter a vertex, which is incorrect.  We're not supposed to finish a vertex until we've finished everything reachable from it.

Reply:
@jkinable @jsichi  Indeed, you need some extra bookkeeping in order to call `finishVertex()` at the correct time. One solution is to add an extra flag on the stack entry to indicate whether this is the first time that this vertex has been added in the stack. When you `pop()` that entry, you can call `finishVertex()`.